### ISSUE TITLE
Failure to terminate the transaction when the broker cannot be connected

### ISSUE THREAD

**Issue by caglargul52** [chkr1011/MQTTnet#1054]
# Failure to terminate the transaction when the broker cannot be connected

Hello.

There is a serious problem with the library.
The scenario is:

Broker : Mosquitto is installed on the docker.

1. I shut down my broker. (Docker Container shutdown)
2. I wrote and ran a console application like the one below.
3. I opened the broker. (Docker Container open)

I subscribed to the broker with MQTTBox and did not receive messages. Also, when I want to close the application, although the application is closed, it appears in the task manager and says Access Denied.

### Testing was done on Windows 10. And MQTTnet 3.0.13

The problem is that we cannot get TcpAck and Conack information from mosquitto because we have closed the docker container directly.

```csharp
static void Main(string[] args)
        {
            IMqttClient mqttClient = new MqttFactory().CreateMqttClient();

            MqttClientOptions mqttOptions = new MqttClientOptions
            {
                ChannelOptions = new MqttClientTcpOptions
                {
                    Server = "localhost", 
                    Port = 1883 // Eg. 9000
                },
                CleanSession = true,
                ClientId = "xXx"
            };

            mqttClient.UseConnectedHandler(e =>
            {
                Console.WriteLine(mqttClient.Options.ClientId + " - bağlandı.");

            });

            mqttClient.UseDisconnectedHandler(e =>
            {
                Console.WriteLine(mqttClient.Options.ClientId + " - brokera bağlantı kesildi yeniden deneniyor...");

                Thread.Sleep(1000);
                mqttClient.ReconnectAsync();
            });


            mqttClient.ConnectAsync(mqttOptions);


            Task.Run(() =>
            {
                while(true)
                {
                    if(mqttClient.IsConnected)
                    {
                        mqttClient.PublishAsync("asd", Encoding.UTF8.GetBytes(Guid.NewGuid().ToString()));
                    }

                    Thread.Sleep(1000);
                }
            });

            Console.Read();
        }
```

**Comment by chkr1011** [comment]
Please attach the trace from the client (activation is described in wiki). Without having this I can only guess what the problem might be. In general the code looks OK to me and works in other projects without issues.

**Comment by caglargul52** [comment]
I found the error in more detail. It works fine when we provide a remote ip address. (Like 10.30.10.55) but when we give localhost or 127.0.0.1 it doesn't work if there is no broker.

**Comment by caglargul52** [comment]
> Please attach the trace from the client (activation is described in wiki). Without having this I can only guess what the problem might be. In general the code looks OK to me and works in other projects without issues.

I found the error in more detail. It works fine when we provide a remote ip address. (Like 10.30.10.55) but when we give localhost or 127.0.0.1 it doesn't work if there is no broker.

**Comment by chkr1011** [comment]
That sounds really strange to me. Because the string is passed directly to the .net socket. 

Which target platform do you use?

**Comment by caglargul52** [comment]
> That sounds really strange to me. Because the string is passed directly to the .net socket.
> 
> Which target platform do you use?

.NET Core 3.1 on Windows 10

**Comment by chkr1011** [comment]
OK then we have to ty some things: Please set DualMode to _false_ in the client settings.

**Comment by chkr1011** [comment]
And please test with latest 3.0.14-rc5.

**Comment by caglargul52** [comment]
> And please test with latest 3.0.14-rc5.

How can I set the dual mode? Also 3.0.14-rc5. where can i download the version?

**Comment by chkr1011** [comment]
And please attach the trace.

DualMode is a property at the client options.

rc5 is available at nuget.

**Comment by caglargul52** [comment]
I installed RC5 and after setting dual mode to false the error was fixed.

Then I went back to 3.0.13. It worked again. It was dual mode.

Thank you...
----------------------------------------------------------------------------------
>> [2021-01-02T11:41:33.4658581Z] [1] [MqttClient] [Verbose]: Trying to connect with server 'localhost:1883' (Timeout=00:00:10).
>> [2021-01-02T11:41:35.7012247Z] [8] [MqttClient] [Error]: Error while connecting with server.
MQTTnet.Exceptions.MqttCommunicationException: Error while connecting with host 'localhost:1883'.
 ---> System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (10061): Hedef makine etkin olarak reddettiğinden bağlantı kurulamadı. [::1]:1883
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw(Exception source)
   at System.Net.Sockets.Socket.EndConnect(IAsyncResult asyncResult)
   at System.Net.Sockets.Socket.DoMultipleAddressConnectCallback(Object result, MultipleAddressConnectAsyncResult context)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw(Exception source)
   at System.Net.Sockets.Socket.DoMultipleAddressConnectCallback(Object result, MultipleAddressConnectAsyncResult context)
   at System.Net.Sockets.Socket.MultipleAddressConnectCallback(IAsyncResult result)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw(Exception source)
   at System.Net.Sockets.Socket.EndConnect(IAsyncResult asyncResult)
   at System.Net.Sockets.Socket.<>c.<ConnectAsync>b__277_0(IAsyncResult iar)
--- End of stack trace from previous location where exception was thrown ---
   at MQTTnet.Implementations.CrossPlatformSocket.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at MQTTnet.Implementations.CrossPlatformSocket.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
   at MQTTnet.Implementations.MqttTcpChannel.ConnectAsync(CancellationToken cancellationToken)
   at MQTTnet.Internal.MqttTaskTimeout.WaitAsync(Func`2 action, TimeSpan timeout, CancellationToken cancellationToken)
   at MQTTnet.Adapter.MqttChannelAdapter.ConnectAsync(TimeSpan timeout, CancellationToken cancellationToken)
   at MQTTnet.Client.MqttClient.ConnectAsync(IMqttClientOptions options, CancellationToken cancellationToken)
>> [2021-01-02T11:41:35.7551017Z] [8] [MqttClient] [Verbose]: Disconnecting [Timeout=00:00:10]
>> [2021-01-02T11:41:35.7560265Z] [8] [MqttClient] [Verbose]: Disconnected from adapter.
>> [2021-01-02T11:41:35.7572828Z] [8] [MqttClient] [Info]: Disconnected.
xXx - brokera bağlantı kesildi yeniden deneniyor...
